src/devices/Mfrc522/Mfrc522Controller.cs

```diff
@@ -0,0 +1,447 @@
+        {
+            WriteSpi((byte)register, value);
+        }
+
+        private void WriteSpi(Register register, Command command)
+        {
+            WriteSpi((byte)register, (byte)command);
+        }
+
+        private byte ReadSpi(byte address)
+        {
+            Span<byte> buffer = stackalloc byte[2] {
+                (byte) (((address << 1) & 0x7E) | 0x80),
+                0
+            };
+            _spiDevice.TransferFullDuplex(buffer, buffer);
+            return buffer[1];
+        }
+
+        private byte ReadSpi(Register register)
+        {
+            return ReadSpi((byte)register);
+        }
+
+        public (Status status, byte[] data) ReadCardData(byte blockAddress)
+        {
+            List<byte> buff = new List<byte>
+            {
+              (byte)RequestMode.Read,
+              blockAddress
+            };
+            var crc = CalulateCRC(buff.ToArray());
+            buff.Add(crc[0]);
+            buff.Add(crc[1]);
+
+            var (status, data, _) = SendCommand(Command.Transceive, buff.ToArray());
+            return (status, data);
+        }
+
+        private void SetBitMask(Register register, byte mask)
+        {
+            var tmp = ReadSpi(register);
+            WriteSpi(register, (byte)(tmp | mask));
+        }
+
+        private void ClearBitMask(Register register, byte mask)
+        {
+            var tmp = ReadSpi(register);
+            WriteSpi(register, (byte)(tmp & (~mask)));
+        }
+
+        private void AntennaOn()
+        {
+            var temp = ReadSpi(TxControlReg);
+            SetBitMask(TxControlReg, 0x03);
+        }
+
+        private void AntennaOff()
+        {
+            ClearBitMask(TxControlReg, 0x03);
+        }
+
+        private (Status status, byte[] backData, byte backLen) SendCommand(Command command, byte[] sendData)
+        {
+            var backData = new List<byte>();
+            byte backLen = 0;
+            var status = Error;
+            byte irqEn = 0x00;
+            byte waitIRq = 0x00;
+            byte n = 0;
+            var i = 0;
+
+            switch(command)
+            {
+                case Command.Authenticate:
+                    irqEn = 0x12;
+                    waitIRq = 0x10;
+                    break;
+
+                case Command.Transceive:
+                    irqEn = 0x77;
+                    waitIRq = 0x30;
+                    break;
+            }
+
+            WriteSpi(CommIEnReg, (byte)(irqEn | 0x80));
+            ClearBitMask(CommIrqReg, 0x80);
+            SetBitMask(FIFOLevelReg, 0x80);
+
+            WriteSpi(CommandReg, (byte)Idle);
+
+            while (i < sendData.Length)
+            {
+                WriteSpi(FIFODataReg, sendData[i]);
+                i++;
+            }
+
+            WriteSpi(CommandReg, (byte)command);
+
+            if (command == Transceive)
+            {
+                SetBitMask(BitFramingReg, 0x80);
+            }
+
+            i = 2000;
```

Comment:
check my changeset I shared with you - you can do this with Stopwatch instead of loop with magic number of iterations :smile: